### PR TITLE
Add k_BT to HPMC integrator

### PR DESCRIPTION
**Pull request by surajkannur**
## Description

Previously, $kT$ was assumed to be 1. In this PR, $kT$ is implemented as a variant in the HPMC integrator.

## Motivation and context

Assuming $kT = 1$ meant that the well depth of the potential had to be rescaled to simulate different temperatures in the case of patchy particles. It also made interpreting the $\beta P$ output of the SDF compute difficult. With this change, 
the temperature can be user-defined.

## How has this been tested?

- [x] pytest set and get kT
- notebook to test 
    - [x] $\mu VT$, 
    - [x] cluster moves, 
    - [x] shape moves
- [x] hoomd-validation tests

hoomd-validation: 
LJ Fluid

LJ Union

Patchy Particle Pressure 

MuVT simulation with octahedron particles initialized at $kT=1.0$ and $f=10$

Patchy particle simulation replicating results of the tutorial: 'Modelling Patchy Particles' with varying kT instead of the energy scale

Effect of kT on elastic shape move acceptance ratio for system adapted from test_shape_updater.py.

TODO documentation

## Change log

```
Add kT to HPMC integrator
```

## Checklist:

- [x] IntegratorHPMCMono.h
- [x] $\mu VT$
- [x] Cluster moves
- [x] SDF pressure compute
- [x] shape moves
- [x] boxMC
- [x] Change SDF input from betaP to P 
- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.rst).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [ ] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

### CODE REVIEW

**Comment by surajkannur** [comment]
pre-commit.ci autofix

File: hoomd/hpmc/ComputeSDF.h [hunk 1]
```diff
@@ -486,6 +486,8 @@ template<class Shape> void ComputeSDF<Shape>::countHistogramLinearSearch(uint64_
     const LongReal min_core_radius = m_mc->getMinCoreDiameter() * LongReal(0.5);
     const auto& pair_energy_search_radius = m_mc->getPairEnergySearchRadius();
 
+    Scalar kT = m_mc->getTimestepkT(timestep); // use const?
```
Comment (joaander) [inline hoomd/hpmc/ComputeSDF.h]:
```suggestion
    const Scalar kT = (*m_mc->getKT())(timestep);
```
I would prefer that we call `operator()` on `Variant` directly.

File: hoomd/hpmc/IntegratorHPMC.cc [hunk 2]
```diff
@@ -258,6 +258,9 @@ void export_IntegratorHPMC(pybind11::module& m)
         .def("getMPS", &IntegratorHPMC::getMPS)
         .def("getCounters", &IntegratorHPMC::getCounters)
         .def("communicate", &IntegratorHPMC::communicate)
+        .def("getkT", &IntegratorHPMC::getkT)
+        .def("setkT", &IntegratorHPMC::setkT)
+        .def_property("kT", &IntegratorHPMC::getkT, &IntegratorHPMC::setkT)
```
Comment (joaander) [inline hoomd/hpmc/IntegratorHPMC.cc]:
```suggestion
        .def_property("kT", &IntegratorHPMC::getKT, &IntegratorHPMC::setKT)
```
We should need only the property.

File: hoomd/hpmc/IntegratorHPMC.h [hunk 3]
```diff
@@ -260,6 +261,31 @@ class PYBIND11_EXPORT IntegratorHPMC : public Integrator
...
+        {
+        m_kT = kT;
+        }
+
+    //! Get kT variant
+    //! \returns current value of kT parameter
+    std::shared_ptr<Variant> getkT()
+        {
+        return m_kT;
+        }
```
Comment (joaander) [inline hoomd/hpmc/IntegratorHPMC.h]:
```suggestion
    std::shared_ptr<const Variant> getKT()
        {
        return m_kT;
        }
```
The capital K is necessary other wise the first word of the method name appears to be `getk` and the second word as `T`.

File: hoomd/hpmc/IntegratorHPMC.h [hunk 4]
```diff
@@ -260,6 +261,31 @@ class PYBIND11_EXPORT IntegratorHPMC : public Integrator
...
+    /** Evaluate the kT variant at the given timestep.
+
+        @param timestep The simulation timestep.
+        @returns The value of the kT variant at the given timestep.
+    */
+    Scalar getTimestepkT(uint64_t timestep)
+        {
+        return m_kT->operator()(timestep);
+        }
+
```
Comment (joaander) [inline hoomd/hpmc/IntegratorHPMC.h]:
```suggestion
```
Not necessary.

File: hoomd/hpmc/IntegratorHPMC.h [hunk 5]
```diff
@@ -260,6 +261,31 @@ class PYBIND11_EXPORT IntegratorHPMC : public Integrator
         return m_nselect;
         }
 
+    //! Set kT variant
+    /*! \param kT new k_BT variant to set
+     */
+    void setkT(std::shared_ptr<Variant> kT)
```
Comment (joaander) [inline hoomd/hpmc/IntegratorHPMC.h]:
```suggestion
    void setKT(std::shared_ptr<Variant> kT)
```

File: hoomd/hpmc/IntegratorHPMCMono.h [hunk 6]
```diff
@@ -948,7 +948,9 @@ void IntegratorHPMCMono<Shape>::update(uint64_t timestep)
                     this->computeOneExternalEnergy(typ_i, pos_i, shape_i.orientation, h_charge.data[i], true);
                 }
 
-            bool accept = !overlap && hoomd::detail::generate_canonical<double>(rng_i) < slow::exp(patch_field_energy_diff);
+            Scalar kT = IntegratorHPMC::getTimestepkT(timestep);
```
Comment (joaander) [inline hoomd/hpmc/IntegratorHPMCMono.h]:
```suggestion
            const Scalar kT = (*m_mc->getKT())(timestep);
```
Also, move this line to the top of the function. `kT` does not vary during the loop, so there is no need to evaluate it every loop iteration.

File: hoomd/hpmc/ShapeMoves.h [hunk 7]
```diff
@@ -472,27 +472,31 @@ class ElasticShapeMove<ShapeConvexPolyhedron> : public ElasticShapeMoveBase<Shap
...
+                                                                                        shape_old,
+                                                                                        iold);
         Scalar stiff = (*m_k)(timestep);
         Matrix3S eps = this->getEps(type_id);
         Matrix3S eps_last = this->getEpsLast(type_id);
         Scalar e_ddot_e = (eps * eps).trace();
         Scalar e_ddot_e_last = (eps_last * eps_last).trace();
-        return N * stiff * (e_ddot_e_last - e_ddot_e) * this->m_volume[type_id] + inertia_term;
+
+        Scalar kT = m_mc->getTimestepkT(timestep);
```
Comment (joaander) [inline hoomd/hpmc/ShapeMoves.h]:
```suggestion
        const Scalar kT = (*m_mc->getKT())(timestep);
```

File: hoomd/hpmc/ShapeMoves.h [hunk 8]
```diff
@@ -472,27 +472,31 @@ class ElasticShapeMove<ShapeConvexPolyhedron> : public ElasticShapeMoveBase<Shap
...
+
+        return (N * stiff * (e_ddot_e_last - e_ddot_e) * this->m_volume[type_id]) / kT
+               + inertia_term - 3 * N * slow::log(1 / kT);
```
Comment (joaander) [inline hoomd/hpmc/ShapeMoves.h]: Why `log(1/kT)` in the last term?

File: hoomd/hpmc/UpdaterBoxMC.cc [hunk 9]
```diff
@@ -281,7 +281,9 @@ inline bool UpdaterBoxMC::box_resize_trial(Scalar Lx,
 
     double p = hoomd::detail::generate_canonical<double>(rng);
 
-    if (allowed && p < exp(-delta_beta_H) * exp(-delta_U_pair) * exp(-delta_U_external))
+    Scalar kT = m_mc->getTimestepkT(timestep);
```
Comment (joaander) [inline hoomd/hpmc/UpdaterBoxMC.cc]:
```suggestion
    const Scalar kT = (*m_mc->getKT())(timestep);
```

File: hoomd/hpmc/UpdaterClusters.h [hunk 10]
```diff
@@ -1694,6 +1694,8 @@ void UpdaterClusters<Shape>::update(uint64_t timestep)
             delta_U[p] = delU;
             }
 
+        Scalar kT = m_mc->getTimestepkT(timestep);
```
Comment (joaander) [inline hoomd/hpmc/UpdaterClusters.h]:
```suggestion
        const Scalar kT = (*m_mc->getKT())(timestep);
```

File: hoomd/hpmc/UpdaterMuVT.h [hunk 11]
```diff
@@ -1093,7 +1093,8 @@ template<class Shape> void UpdaterMuVT<Shape>::update(uint64_t timestep)
 
                 if (nonzero)
                     {
-                    lnboltzmann += lnb;
+                    Scalar kT = m_mc->getTimestepkT(timestep);
```
Comment (joaander) [inline hoomd/hpmc/UpdaterMuVT.h]:
```suggestion
                    const Scalar kT = (*m_mc->getKT())(timestep);
```
Move this line to the top of the function. `kT` does not need to be reevaluated in the inner loops.

File: hoomd/hpmc/UpdaterMuVT.h [hunk 12]
```diff
@@ -1274,7 +1275,8 @@ template<class Shape> void UpdaterMuVT<Shape>::update(uint64_t timestep)
             Scalar lnb(0.0);
             if (tryRemoveParticle(timestep, tag, lnb))
                 {
-                lnboltzmann += lnb;
+                Scalar kT = m_mc->getTimestepkT(timestep);
```
Comment (joaander) [inline hoomd/hpmc/UpdaterMuVT.h]:
```suggestion
```
Not needed when `kT` is evaluated at the top of the function.

File: hoomd/hpmc/UpdaterMuVT.h [hunk 13]
```diff
@@ -1534,9 +1536,10 @@ template<class Shape> void UpdaterMuVT<Shape>::update(uint64_t timestep)
                          &stat);
 
                 // apply criterion on rank zero
+                Scalar kT = m_mc->getTimestepkT(timestep);
```
Comment (joaander) [inline hoomd/hpmc/UpdaterMuVT.h]:
```suggestion
```
Not needed when `kT` is evaluated at the top of the function.

File: hoomd/hpmc/UpdaterBoxMC.cc [hunk 14]
```diff
@@ -281,7 +281,9 @@ inline bool UpdaterBoxMC::box_resize_trial(Scalar Lx,
...
+
+    if (allowed && p < exp(-(delta_beta_H + delta_U_pair + delta_U_external) / kT))
```
Comment (joaander) [inline hoomd/hpmc/UpdaterBoxMC.cc]: This is incorrect and should appear as a failure in validation tests when `kT != 1.0`. The variable named `delta_beta_H` is not what it is named. See lines 501 and 584 - it includes the `log(V_new/V)` term as well. This term should not divided by `kT` (see Frenkel and Smit or https://cassandra-mc.readthedocs.io/en/latest/theory/theory.html#volume-scaling)

To fix, modify `box_resize_trial` to accept an additional parameter: `log_V_term`. Then modify the code around lines 501 and 584 to compute this term and pass it into `box_resize_trial` separately.

**Review by joaander — CHANGES_REQUESTED** [review]
Thanks! I have made several suggestions to improve code style. I also found a bug in `BoxMC` (not your fault, the variable is named incorrectly).

As an added TODO item, we need to modify `hpmc/compute.py`.
- [x] Add `SDF.P` which returns `SDF.betaP * {the current kT}`.

I also considered renaming `betaP` to `P`, but I now think we should keep `betaP` for backwards compatibility. `betaP` is also more natural for purely hard particle systems.

File: hoomd/hpmc/UpdaterBoxMC.cc [hunk 15]
```diff
@@ -284,7 +284,7 @@ inline bool UpdaterBoxMC::box_resize_trial(Scalar Lx,
 
     const Scalar kT = (*m_mc->getKT())(timestep);
 
-    if (allowed && p < exp(-(delta_beta_H + delta_U_pair + delta_U_external) / kT) + log_V_term)
```
Comment (joaander) [inline hoomd/hpmc/UpdaterBoxMC.cc]: Are you sure this change is correct? The cassandra docs show $P\\Delta V$ (which HOOMD calls `delta_beta_H`) multiplied by $\beta$. You can check in Frenkel and Smit to verify.

If you found errors in testing, double check the sign on the `log_V_term`.

**Review by joaander — APPROVED** [review]
Thanks!

**Pull request merged**